core/src/server/rpc_module.rs

```diff
@@ -721,39 +561,73 @@ impl<Context: Send + Sync + 'static> RpcModule<Context> {
 	///
 	/// # Returns
 	///
-	/// An async block which returns `Result<(), SubscriptionCallbackError>` the error is simply
-	/// for a more ergonomic API and is not used (except logged for user-related caused errors).
-	/// By default jsonrpsee doesn't send any special close notification,
-	/// it can be a footgun if one wants to send out a "special notification" to indicate that an error occurred.
+	/// An async block which returns something that implements [`crate::server::IntoSubscriptionResponse`] which
+	/// decides what action to take once the subscription callback returns i.e, whether to sent out another message
+	/// on the subscription stream before closing down it.
+	///
+	/// This is implemented for `Result<T, E>` and `Option<T>`.
+	///
+	/// It's recommended to use `Result` if you want to propagate the error as special error notification
+	/// and `Option` if no error's should be propagated.
 	///
-	/// If you want to a special error notification use `SubscriptionSink::close` or
-	/// `SubscriptionSink::send` before returning from the async block.
+	/// Another option is to implement [`crate::server::IntoSubscriptionResponse`] if you want customized behaviour.
+	///
+	/// The error notification has the following format:
+	///
+	/// ```json
+	/// {
+	///  "jsonrpc": "2.0",
+	///  "method": "<method>",
+	///  "params": {
+	///    "subscription": "<subscriptionID>",
+	///    "error": <your msg>
+	///    }
+	///  }
+	/// }
+	/// ```
 	///
 	/// # Examples
 	///
 	/// ```no_run
 	///
-	/// use jsonrpsee_core::server::rpc_module::{RpcModule, SubscriptionSink, SubscriptionMessage};
+	/// use jsonrpsee_core::server::{RpcModule, SubscriptionSink, SubscriptionMessage};
 	/// use jsonrpsee_core::Error;
+	/// use jsonrpsee_types::ErrorObjectOwned;
 	///
 	/// let mut ctx = RpcModule::new(99_usize);
 	/// ctx.register_subscription("sub", "notif_name", "unsub", |params, pending, ctx| async move {
-	///     let x = params.one::<usize>()?;
 	///
-	///     // mark the subscription is accepted after the params has been parsed successful.
+	///     let x = match params.one::<usize>() {
+	///         Ok(x) => x,
+	///         Err(e) => {
+	///            pending.reject(ErrorObjectOwned::from(e)).await;
```

Comment:
I wondered whether a user could return an error for this too, but realised that you _have_ to return some sort of error for this, so yeah I see why it's separate and it makes sense to me!

(I guess 99% of the time people will just `pending.accept()` anyway!)